plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java

```diff
@@ -0,0 +1,614 @@
+      component.setProperty(StorageFacet.P_FORMAT, getRepository().getFormat().getValue()); // M1 or M2!!!
+      component.setProperty(StorageFacet.P_PATH, path.getPath());
+      component.setProperty(StorageFacet.P_GROUP, path.getGroupId());
+      component.setProperty(StorageFacet.P_NAME, path.getArtifactId());
+      component.setProperty(StorageFacet.P_VERSION, path.getVersion());
+      component.setProperty(StorageFacet.P_LAST_UPDATED, content.getLastUpdated().toDate());
+
+      final NestedAttributesMap componentAttributes = getFormatAttributes(tx, component);
+      componentAttributes.set(P_COMPONENT_KEY, getComponentKey(path));
+      componentAttributes.set(P_GROUP_ID, path.getGroupId());
+      componentAttributes.set(P_ARTIFACT_ID, path.getArtifactId());
+      componentAttributes.set(P_VERSION, path.getVersion());
+      if (path.isSnapshot()) {
+        componentAttributes.set(P_BASE_VERSION, path.getBaseVersion());
+      }
+    }
+
+    OrientVertex asset = selectComponentAsset(tx, component, path);
+    if (asset == null) {
+      asset = tx.createAsset(tx.getBucket());
+      asset.addEdge(StorageFacet.E_PART_OF_COMPONENT, component);
+      asset.setProperty(StorageFacet.P_FORMAT, getRepository().getFormat().getValue());
+      asset.setProperty(StorageFacet.P_PATH, path.getPath());
+      asset.setProperty(StorageFacet.P_NAME, path.getFileName());
+
+      final NestedAttributesMap assetAttributes = getFormatAttributes(tx, asset);
+      assetAttributes.set(P_ASSET_KEY, getAssetKey(path));
+      assetAttributes.set(P_ASSET_KIND, ASSET_KIND_ARTIFACT);
+      assetAttributes.set(P_GROUP_ID, path.getGroupId());
+      assetAttributes.set(P_ARTIFACT_ID, path.getArtifactId());
+      assetAttributes.set(P_VERSION, path.getVersion());
+      if (path.isSnapshot()) {
+        assetAttributes.set(P_BASE_VERSION, path.getBaseVersion());
+      }
+      if (path.getClassifier() != null) {
+        assetAttributes.set(P_CLASSIFIER, path.getClassifier());
+      }
+      assetAttributes.set(P_EXTENSION, path.getExtension());
+
+      // TODO: if subordinate asset (sha1/md5), should we link it somehow to main asset?
+    }
+
+    storeAssetBlob(tx, asset, path, content);
+
+    if (path.isHash()) {
+      verifyAssetChecksum(tx, path);
+    }
+    else {
+      if (!content.getHashAlgorithms().isEmpty()) {
+        for (HashType hashType : HashType.values()) {
+          final HashCode hashCode = content.getHashCode(hashType.getHashAlgorithm());
+          if (hashCode == null) {
+            continue;
+          }
+          putArtifactTx(
+              tx,
+              path.hash(hashType),
+              new MavenPayload(
+                  new StringPayload(hashCode.toString(), "text/plain"),
+                  content.getLastUpdated(),
+                  content.getLastModified(),
+                  Collections.<HashAlgorithm, HashCode>emptyMap()
+              )
+          );
+        }
+      }
+    }
```

Comment:
This code executes only when `!content.getHashAlgorithms().isEmpty()`, and they are always empty in case of hosted repository. The hashes on payload are set on proxy repository (other PR), where client fetches content AND hashes preemptively, before client even asked for them.